### PR TITLE
New Goodie: Ajax Frameworks

### PR DESCRIPTION
**Pull request by sahildua2305**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [x] New Instant Answer
    - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
    - [x] Other: **`New Ajax Frameworks Instant Answer`**
- [ ] Improvement
    - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
    - [ ] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{GoodieRole/Templates/Tests/Docs}: {Short Description}`**

###### People to notify (@mention interested parties)

---

Instant Answer Page: https://duck.co/ia/view/ajax_frameworks

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @mention

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by sahildua2305** [comment]
@moollaza Please deploy this on beta.

**Comment by GuiltyDolphin** [comment]
@sahildua2305 This looks awfully like a cheat sheet w/ the `link` template; is there anything you can do to 'Spice' this up a little? Perhaps an API that has lists that can adapt to change?

Currently this just looks like the information from the list on Wikipedia - is there perhaps some additional information we can provide? E.g., an overview of each of the frameworks? Allowing users to search for a particular framework and receive in-depth information on that framework?

**Comment by sahildua2305** [comment]
@GuiltyDolphin Actually it doesn't just show information from one Wikipedia page. It actually shows names of all famous frameworks from one page and then shows the relevant links for each of the frameworks from their corresponding wiki pages.

The aim of this IA is to provide a quick access to these famous frameworks so that users don't have to keep looking on multiple wiki pages or don't have to search for framework name manually.

Also, the `Spicing` up is a good idea, however I don't think there is any API for this. This is better as a static information as it is less likely to change frequently.

File: t/AjaxFrameworks.t [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            title => "Framework for implementation of Ajax using 'python'",
+            list => [
+                {
+                    name => "Pyjs (Pyjamas)",
+                    link => "http://pyjs.org/"
+                }
+            ]
+        },
+        templates => $templates,
+        meta => $meta
```
Comment (sahildua2305) [inline t/AjaxFrameworks.t]: @GuiltyDolphin Good catch! I will update that.

**Comment by GuiltyDolphin** [comment]
@sahildua2305 I don't think it would be a bad experience for users; the content is short enough that we could easily display all the information w/o cluttering.

Have you thought of any ideas that could make this Goodie *wow* users? What additional information do you think they'd be looking for when searching for this kind of query; how can we incorporate this? Are there some criteria by which we can rank these frameworks?

**Comment by sahildua2305** [comment]
Ok, I will dig deeper to see what other information can be useful for users searching for such queries. @GuiltyDolphin

**Comment by sahildua2305** [comment]
@GuiltyDolphin I think showing size of each of the frameworks can be a good piece of information to show. What do you think?

**Comment by GuiltyDolphin** [comment]
@sahildua2305 Can you elaborate on what you mean by 'size'? Community? Functionality?

**Comment by sahildua2305** [comment]
@GuiltyDolphin I am working on including more information about these frameworks which might be useful for users. I'll push in a while.

**Comment by sahildua2305** [comment]
@GuiltyDolphin Please check now. I have added a lot of more information. It was a very tiring job finding all relevant links for all frameworks, but will be worth the effort if it's useful for the users.

As of now, the frameworks are sorted according to their popularity according to Wikipedia.

JavaScript frameworks:
selection_150

Java frameworks:
selection_148

@moollaza @gaulrobe @pjhampton feedback please :)

**Comment by GuiltyDolphin** [comment]
@sahildua2305 Awesome - that's great; definitely worth its salt as a Goodie now :heart:

@moollaza Wouldn't mind having a poke at this on Beta!

File: lib/DDG/Goodie/AjaxFrameworks.pm [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    my ($lang, @lang_frameworks, @frameworks_list, $title);
+
+    # Return if the remainder doesn't contain the keyword 'framework'
+    return unless $_ =~ /framework/i;
+
+    # Remove keywords 'framework' and 'frameworks' from the remainder
+    #
+    # Useful for queries like 'ajax javascript frameworks'
+    # Remainder in such cases is 'javascript frameworks'
+    $_ =~ s/(frameworks|framework)//gi;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/AjaxFrameworks.pm]: `s/frameworks?//gi`

File: t/AjaxFrameworks.t [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        data => ignore(),
+        templates => $templates,
+        meta => $meta
+    }
+);
+
+my @java_frameworks = (
+    "Frameworks for implementation of Ajax using 'java'",
+    structured_answer => {
+        data => ignore(),
```
Comment (GuiltyDolphin) [inline t/AjaxFrameworks.t]: We can use `Test::Deep`'s routines to test aspects of the response without needing to check the entire structure, e.g., `data => superhashof({ title => re(qr/python/) })` to check that the title contains 'python'.

**Comment by GuiltyDolphin** [comment]
@moollaza Will also want to get some designy eyes on this.

**Comment by ManrajGrover** [comment]
@sahildua2305 @GuiltyDolphin I feel description of frameworks mentioned would be more helpful :)

**Comment by GuiltyDolphin** [comment]
@sahildua2305 Hmm, we have quite a lot of space on the 'size' section, perhaps we could have a description there? I agree with @ManrajGrover - a summary of the framework sounds like it would be quite useful.

**Comment by GuiltyDolphin** [comment]
@sahildua2305 I'd rather get some eyes on this from design before we start doing anything drastic.

cc @moollaza

**Comment by sahildua2305** [comment]
@moollaza please provide some feedback on this one too.

**Comment by moollaza** [comment]
Spoke to @sahildua2305 about this -- I would prefer to see the data maintained outside of DuckDuckGo (e.g. a separate repo), and pulled in via JSON to power a Spice IA. 

This way the IA code and data, which is subject to opinions and frequent change are separate.

**Pull request closed without merging**